### PR TITLE
Support Navigation

### PR DESCRIPTION
**Pull request by Patrick-Kladek**
### Description

Allows to use maplibre-navigation-ios with SwiftUI without duplicating targets or source code. We think this is the best way given the following requirements:

* No dependency on maplibre-navigation-ios, so users without navigation are not increasing their app size.
* Easy maintenance as changes are affecting both (maplibre & maplibre-navigation), eg. camera improvements

### Tasks

- [x] duplicate MapView to NavigationMapView
- [x] add binding to a route and start/stopp navigation when route is available
- [x] extract common requirements in protocol
- [x] remove maplibre-navigation dependency
- [x] extract start/stop logic to consumer and test in existing app

### Infos for Reviewer

Note #37 should be merged first as this PR contains the changes already.

Relevant changes are in files:

* `MapView.swift`
* `MapViewController.swift`
* `MapViewCoordinator.swift`
* `MapViewModifiers.swift`

To make this work we are not using `UIViewRepresentable`, instead we are using `UIViewControllerRepresentable` and providing a default ViewController that has a mapView. This requirement is extracted into a protocol `WrappedViewController` and allows to instantiate a MapView specifying the ViewController via Generics:

```swift
@ViewBuilder
var mapView: some View {
    MapView<MapViewController>(styleURL: self.styleURL, camera: self.$camera) {
...
    }
}
```

Thats the only change needed if you want to continue using this Package without navigation support. However if you want to support navigation the following changes are needed:

Add this branch as a dependency to your project: https://github.com/maplibre/maplibre-navigation-ios/pull/54

```swift
import MapboxCoreNavigation
import MapboxNavigation

extension NavigationViewController: WrappedViewController {
    public typealias MapType = NavigationMapView
}


@State var route: Route?
@State var navigationInProgress: Bool = false

@ViewBuilder
var mapView: some View {
    MapView<NavigationViewController>(makeViewController: NavigationViewController(dayStyleURL: self.styleURL), styleURL: self.styleURL, camera: self.$mapStore.camera) {

    }
    .unsafeMapViewControllerModifier { navigationViewController in
        navigationViewController.delegate = self.mapStore
        if let route = self.route, self.navigationInProgress == false {
            let locationManager = SimulatedLocationManager(route: route)
            navigationViewController.startNavigation(with: route, locationManager: locationManager)
            self.navigationInProgress = true
        } else if self.route == nil, self.navigationInProgress == true {
            navigationViewController.endNavigation()
            self.navigationInProgress = false
        }

        navigationViewController.mapView.showsUserLocation = self.showUserLocation && self.mapStore.streetView == .disabled
    }
    .cameraModifierDisabled(self.route != nil)
}
```

### CODE REVIEW

**Comment by ianthetechie** [comment]
Looks like there's a merge conflict now. Could you fix that real quick for me make it easier to review? GitHub appears to just throw up its hands and gives a useless diff now 😂

**Comment by Archdoog** [comment]
Thanks for the PR @Patrick-Kladek!

Taking a look over this concept, it looks like the generalized goal is to introduce the add the convenience of maplibre/mapbox’s NavigationViewController to the SwiftUI MapView?

While there’s no harm in this approach, I think it’s a specialization that doesn’t align with the goals of this project. I’d propose this alternative:

1. Create the navigation route style needed using the Symbol DSL in this repo. Here’s how we did it in ferrostar [RouteStyleLayer.swift](https://github.com/stadiamaps/ferrostar/blob/main/apple/Sources/FerrostarMapLibreUI/Layers/RouteStyleLayer.swift).
2. That way you can just display the route like any other symbol on the map based on what the navigation repo has as the current route in state. This could easily be a @State driven variable from the navigation controller delegate updates.
3. At this point, you could easily add the SwiftUI MapView to the maplibre-navigation UIKit implementation, replacing the `NavigationMapView`. OR better, why not just drive a full SwiftUI view w/ the navigation repo’s controller/core state, wrapping whatever UIKit stuff you need in their own simple representables?

I realize this may not be as convenient as just swapping out the MLNMapView, but having used Mapbox’s navigation without the full view controller in my own iOS app, I think it would actually be relatively reasonable to follow that process, even if it was just in your app implementation.

Feel free to reach out over Slack, happy to dive in, discuss in more detail and provide some examples of how I've done the above myself before.

**Comment by Patrick-Kladek** [comment]
@ianthetechie just fixed the merge conflicts

**Comment by hactar** [comment]
> Thanks for the PR @Patrick-Kladek!
> 
> Taking a look over this concept, it looks like the generalized goal is to introduce the add the convenience of maplibre/mapbox’s NavigationViewController to the SwiftUI MapView?
> 
> While there’s no harm in this approach, I think it’s a specialization that doesn’t align with the goals of this project. I’d propose this alternative:
> 
> 1. Create the navigation route style needed using the Symbol DSL in this repo. Here’s how we did it in ferrostar [RouteStyleLayer.swift](https://github.com/stadiamaps/ferrostar/blob/main/apple/Sources/FerrostarMapLibreUI/Layers/RouteStyleLayer.swift).
> 2. That way you can just display the route like any other symbol on the map based on what the navigation repo has as the current route in state. This could easily be a @State driven variable from the navigation controller delegate updates.
> 3. At this point, you could easily add the SwiftUI MapView to the maplibre-navigation UIKit implementation, replacing the `NavigationMapView`. OR better, why not just drive a full SwiftUI view w/ the navigation repo’s controller/core state, wrapping whatever UIKit stuff you need in their own simple representables?
> 
> I realize this may not be as convenient as just swapping out the MLNMapView, but having used Mapbox’s navigation without the full view controller in my own iOS app, I think it would actually be relatively reasonable to follow that process, even if it was just in your app implementation.
> 
> Feel free to reach out over Slack, happy to dive in, discuss in more detail and provide some examples of how I've done the above myself before.

I have adjusted the PR to include a convenience init so that the call site no longer changes if you do not need a UIViewController backing, so the old `MapView(...)` init returns, which I think is much cleaner and I hope resolves some potential reservations.

As for aligning with project goals: I do agree that a specialization towards MapLibre Navigation is not a project goal. But I don't think this is was this PR is.

In this revised PR, the call site for all existing users does not change - for anyone who does not need a UIViewController based backing, nothing changes. But for anyone who does, for example to include a UIViewController based navigation system such as MapLibre Navigation, this PR opens up this possibility. This is also not the only reason someone might need a UIViewController backed system: UIViewController offers all that UIView does, plus more, and UIViewRepresentable is just a convenience for UIViewControllerRespresentable I would argue - which is more powerful.

In other words, personally I do not see this PR as a specialization, but as a generalization, and now swiftly hidden behind a convenience init :). 

We thought of some other alternatives and tried discussing them here: https://osmus.slack.com/archives/C06U5MM097B/p1717157153623649

Maybe we can discuss this in one of the calls on Wednesday? ☺️

File: Sources/MapLibreSwiftUI/MapViewModifiers.swift [hunk 1]
```diff
@@ -141,4 +135,10 @@ public extension MapView {
         result.onViewPortChanged = onViewPortChanged
         return result
     }
+
+    func cameraModifierDisabled(_ disabled: Bool) -> Self {
```
Comment (Archdoog) [inline Sources/MapLibreSwiftUI/MapViewModifiers.swift:139]: Can we add a docstring explaining what this handles?

File: Sources/MapLibreSwiftUI/MapView.swift [hunk 2]
```diff
@@ -31,96 +37,118 @@ public struct MapView: UIViewRepresentable {
...
+        context.coordinator.updateLayers(mapView: uiViewController.mapView)
 
         // FIXME: This isn't exactly telling us if the *map* is loaded, and the docs for setCenter say it needs to be.
-        let isStyleLoaded = mapView.style != nil
+        let isStyleLoaded = uiViewController.mapView.style != nil
 
-        context.coordinator.updateCamera(mapView: mapView,
-                                         camera: $camera.wrappedValue,
-                                         animated: isStyleLoaded)
+        if cameraDisabled == false {
```
Comment (Archdoog) [inline Sources/MapLibreSwiftUI/MapView.swift:108]: Can you clarify the purpose/use case for this? We spent some good time enabling all of the potential use cases the underlying MLNMapView offers through the MapViewCamera, so I just want make sure there's not a way to already accomplish the same end goal. Or a bug/use case were still missing with the camera.

**Review by Archdoog — COMMENTED** [review]
Generally looking pretty good 👍. Have a couple of questions and suggestions.

File: Sources/MapLibreSwiftUI/MapView.swift (hunk 2 above)
Comment (Patrick-Kladek) [reply Sources/MapLibreSwiftUI/MapView.swift:108]: When Maplibre-navigation is running, it is in control of the camera. We want to prevent Maplibre-DSL from interfering with the camera, that's why we added `cameraModifierDisabled()`

Comment (ianthetechie) [reply Sources/MapLibreSwiftUI/MapView.swift:108]: Makes sense to me. Thanks!

**Review by ianthetechie — APPROVED** [review]
